domain/network/state/linklayer.go

```diff
@@ -123,6 +123,201 @@ func (st *State) SetMachineNetConfig(ctx context.Context, nodeUUID string, nics
+// any operation fails during execution.
+func (st *State) GetAllLinkLayerDevicesByNetNodeUUIDs(ctx context.Context) (map[string][]network.NetInterface, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var llds []getLinkLayerDevice
+	var dnsDomains []dnsSearchDomainRow
+	var dnsAddresses []dnsAddressRow
+	var ipAddresses []getIpAddress
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		llds, err = st.getAllLinkLayerDevices(ctx, tx)
+		if err != nil {
+			return errors.Errorf("fetching all link layer devices: %w", err)
+		}
+		dnsDomains, err = st.getAllDNSDomains(ctx, tx)
+		if err != nil {
+			return errors.Errorf("fetching all DNS search domains: %w", err)
+		}
+		dnsAddresses, err = st.getAllDNSAddresses(ctx, tx)
+		if err != nil {
+			return errors.Errorf("fetching all DNS addresses: %w", err)
+		}
+		ipAddresses, err = st.getAllAddresses(ctx, tx)
+		if err != nil {
+			return errors.Errorf("fetching all IP addresses: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Errorf("fetching all link layer devices: %w", err)
+	}
+
+	dnsDomainByDeviceUUID, _ := accumulateToMap(dnsDomains, func(in dnsSearchDomainRow) (string, string, error) {
+		return in.DeviceUUID, in.SearchDomain, nil
+	})
+	dnsAddressesByDeviceUUID, _ := accumulateToMap(dnsAddresses, func(in dnsAddressRow) (string, string, error) {
+		return in.DeviceUUID, in.Address, nil
+	})
+	ipAddressByDeviceUUID, _ := accumulateToMap(ipAddresses, func(f getIpAddress) (string, getIpAddress, error) {
+		return f.DeviceUUID, f, nil
+	})
+
+	return accumulateToMap(llds, func(in getLinkLayerDevice) (string, network.NetInterface, error) {
+		result, err := dmlToNetInterface(in,
+			dnsDomainByDeviceUUID[in.UUID],
+			dnsAddressesByDeviceUUID[in.UUID],
+			ipAddressByDeviceUUID[in.UUID])
+		return in.NetNodeUUID, result, err
+	})
+}
+
+// accumulateToMap transforms a slice of elements into a map of keys to slices
+// of values using the provided transform function.
+// If the transformation function results in an error, end the loop and return
+// the error
+func accumulateToMap[F any, K comparable, V any](from []F, transform func(F) (K, V, error)) (map[K][]V, error) {
+	to := make(map[K][]V)
+	for _, oneFrom := range from {
+		k, v, err := transform(oneFrom)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		to[k] = append(to[k], v)
+	}
+	return to, nil
+}
+
+// getAllLinkLayerDevices fetches all link-layer devices from the database
+// within the context of a transaction.
+// It executes a SQL query to retrieve device fields, including UUID, name,
+// provider details, and other attributes.
+// The method returns a slice of getLinkLayerDevice and an error if the
+// query preparation or execution fails.
+func (st *State) getAllLinkLayerDevices(ctx context.Context, tx *sqlair.TX) ([]getLinkLayerDevice, error) {
+	stmt, err := st.Prepare(`
+SELECT 
```

Comment:
There is a view that you can use to get everything at once, which fits in `machineInterfaceRow`. See if you can use that and modify as required rather than doing the queries separately.